### PR TITLE
mm/iob: Simplify IOB alloc/free logic

### PR DESCRIPTION
**Pull request by tmedicci**
## Summary

- `g_iob_sem.semcount` must be equal to the total number of free IOBs. It can also be negative if there are no free IOBs and there are threads
waiting for an IOB.
- g_throttle_sem.semcount represents the number of IOBs available for throttled IOB allocations. Like any other semaphore, it should only go
negative if there is a thread waiting for it.
- Both semaphores are related to the same resource (free IOBs), hence, they must be incremented/decremented simultaneously:
  - Whenever a IOB buffer is freed, if a thread is waiting for a non-throttled IOB or a thread is waiting for a throttled IOB and we have at least `CONFIG_IOB_THROTTLE` buffers available, the IOB is put in the committed list (`g_iob_committed`). Otherwise, it is put in the
common free list (`g_iob_freelist`).
  - `g_iob_sem` is always incremented when an IOB buffer is freed, but `g_throttle_sem` is incremented only if we have at least `CONFIG_IOB_THROTTLE` buffers free.
  - Both semaphores are posted with the schedule locked to avoid any mismatches in the semaphores count. This avoids `DEBUGASSERT(g_throttle_sem.semcount <= (CONFIG_IOB_NBUFFERS - CONFIG_IOB_THROTTLE))` asserting after a few hours of intensive use (tested with `esp32-devkitc:wifi_smp` with `iperf` in client mode).
  - If a task is waiting for an IOB semaphore (`iob_allocwait`) is awakened and would check the `g_iob_committed`. The highest priority task waiting for a semaphore will be awakened first.

## Impact

Make it easier to understand and deal with IOB buffer management, avoid `DEBUGASSERT(g_throttle_sem.semcount <= (CONFIG_IOB_NBUFFERS - CONFIG_IOB_THROTTLE))` asserting and implement suggestions made in https://github.com/apache/nuttx/pull/9741#issuecomment-1627821161 and https://github.com/apache/nuttx/pull/9741#issuecomment-1627828588

## Testing

Internal CI testing + ESP32-DevKitC V4 (`esp32-devkitc:wifi_smp`) and `rv-virt:netnsh_smp` long-term duration tests with `iperf` (`perf -c <server_ip> -p 11111 -i 1`). Performance is not affected by this change:
- In `esp32-devkitc:wifi_smp` iperf test (debug assertions enabled), performance even increased (median 8.21Mbits/sec with the patch vs 5.97Mbits/sec without);
- In `rv-virt:netnsh_smp`, the performance had no significant change (median 329.30Mbits/sec with the patch vs 335.3Mbits/sec without)

### CODE REVIEW

**Review by acassis — APPROVED** [review]

File: mm/iob/iob_free.c [hunk 1]
```diff
@@ -151,21 +158,30 @@ FAR struct iob_s *iob_free(FAR struct iob_s *iob)
...
-   * owns an IOB and should find it in the committed list.
+  /* Signal that an IOB is available. This is done with schedule locked
+   * to make sure that both g_iob_sem and g_throttle_sem are incremented
+   * together (if applicable). After the schedule is unlocked, if there
+   * is a thread blocked, waiting for an IOB, this will wake up exactly
+   * one thread.  The semaphore count will correctly indicate that the
+   * awakened task owns an IOB and should find it in the committed list.
    */
 
+  sched_lock();
```
Comment (xiaoxiang781216) [inline mm/iob/iob_free.c:169]: how to handle SMP case?

File: mm/iob/iob_alloc.c [hunk 2]
```diff
@@ -163,36 +172,6 @@ static FAR struct iob_s *iob_allocwait(bool throttled, unsigned int timeout)
...
           iob = iob_alloc_committed();
-          if (iob == NULL)
-            {
-              /* We need release our count so that it is available to
-               * iob_tryalloc(), perhaps allowing another thread to take our
-               * count.  In that event, iob_tryalloc() will fail above and
-               * we will have to wait again.
-               */
-
-              sem->semcount++;
```
Comment (xiaoxiang781216) [inline mm/iob/iob_alloc.c:174]: why skip increasement

File: mm/iob/iob_free.c (hunk 1 above)
Comment (tmedicci) [reply mm/iob/iob_free.c:169]: `sched_lock` has its own logic to handle with the SMP case. In fact, `DEBUGASSERT(g_throttle_sem.semcount <= (CONFIG_IOB_NBUFFERS - CONFIG_IOB_THROTTLE))` was asserting when SMP was enabled (running `iperf` for a few minutes). There was a mismatch between the counters (as posting to the `g_iob_sem` semaphore may switch the context before updating the `g_throttle_sem` value). Locking the scheduler fixed the mismatch...

File: mm/iob/iob_alloc.c (hunk 2 above)
Comment (tmedicci) [reply mm/iob/iob_alloc.c:174]: As soon as an IOB buffer is freed, it becomes available in `iob_alloc_committed` if a task is waiting for a throttled or non-throttled IOB. We eliminated the need to increase the semaphore momentaneously and calling `tryalloc`: IOBs in  `g_iob_committed` do not rely on semaphores values to be allocated. They were already incremented by `nxsem_post` and just acquired by the waiting task by calling `iob_alloc_committed`.

### Example: Waiting for a throttled IOB

#### Before this commit:
- A task is waiting for a throttled IOB buffer but we have non-throttled IOB buffers available (`g_iob_buffer` is positive and `g_throttled_sem` negative).
- An IOB is just freed: it will be put in the free list (`g_iob_freelist`) instead of the committed list.
- The task waiting for the throttled IOB will be awakened by posting to the `g_throttled_sem` (that may become equal to `0`, if only a single thread is waiting).
- We need to temporarily increase the `g_throttled_sem` to be able to alloc the throttled IOB in `iob_tryalloc`.
 
This has some flaws: we may suffer with a throttled/non-throttled mismatch in the SMP case, for instance, and `tryalloc` may fail and we had already incremented the semaphore value (which would not be decremented properly).

#### Current implementation:

We avoid making any "single" manipulations with semaphore values. They are only incremented and decrement together. Considering the same example:
- As soon as the IOB is freed and we have a task waiting for it (either throttled or non-throttled), the IOB is put in the committed list.
- Both semaphores values are incremented by `nxsem_post`.
- If committed, at least one task will be awakened. It will acquire the IOB from the committed list and decrement it in `iob_alloc_committed`. The semaphore counter means the amount of free or committed semaphores.

File: mm/iob/iob_free.c (hunk 1 above)
Comment (tmedicci) [reply mm/iob/iob_free.c:169]: A better explanation about it is in the following comment: https://github.com/apache/nuttx/pull/12501#discussion_r1641875447

File: mm/iob/iob_alloc.c (hunk 2 above)
Comment (tmedicci) [reply mm/iob/iob_alloc.c:174]: @xiaoxiang781216, did this explanation make sense for you? Anything else I can contribute to it?

Comment (xiaoxiang781216) [reply mm/iob/iob_alloc.c:174]: Sorry, for later response, I will review today.

Comment (tmedicci) [reply mm/iob/iob_alloc.c:174]: @xiaoxiang781216 , were you able to take a look?

Comment (tmedicci) [reply mm/iob/iob_alloc.c:174]: @wengzhe , can you please take a look too?

File: mm/iob/iob_alloc.c [hunk 3]
```diff
@@ -87,6 +87,15 @@ static FAR struct iob_s *iob_alloc_committed(void)
 
       g_iob_committed = iob->io_flink;
 
+      g_iob_sem.semcount--;
+
+#if CONFIG_IOB_THROTTLE > 0
+      if (g_throttle_sem.semcount > 0)
```
Comment (wengzhe) [inline mm/iob/iob_alloc.c:93]: - Question 1: How could `g_throttle_sem.semcount` be greater than 0 here?
- Question 2: I want to make sure you've considered these situations, which may relate to these codes:
  - Situation 1: `iob_alloc_committed` for **throttled** IOBs
    - Starting with `g_iob_sem.semcount = CONFIG_IOB_THROTTLE` and `g_throttle_sem.semcount = 0`
    - Task A: `iob_allocwait` with `throttled=true`, so `g_throttle_sem` is decreased by `nxsem_wait_uninterruptible` (now `g_throttle_sem.semcount = -1`)
    - Task B: `iob_free`, which posts to both `g_iob_sem` and `g_throttle_sem` (now `g_iob_sem.semcount = CONFIG_IOB_THROTTLE + 1` and `g_throttle_sem.semcount = 0`)
    - So I think we don't need to decrease `g_throttle_sem` here under this situation (yes we don't, because `g_throttle_sem.semcount` won't be greater than 0 here, and maybe never?), but we need to decrease `g_iob_sem`.
  - Situation 2: `iob_alloc_committed` for **non-throttled** IOBs
    - Starting with `g_iob_sem.semcount = 0`
    - Task A: `iob_allocwait` with `throttled=false`, so `g_iob_sem` is decreased by `nxsem_wait_uninterruptible` (now `g_iob_sem.semcount = -1`)
    - Task B: `iob_free`, which posts only to `g_iob_sem` (now `g_iob_sem.semcount = 0`)
    - Then shall we still decrease the sem on L90 under this case?

Comment (tmedicci) [reply mm/iob/iob_alloc.c]: Hi @wengzhe 

I'm sorry for the late response. There were some implementation flaws in these corner cases. In fact, we just need to decrement `g_iob_sem.semcount` when we have added the just-freed IOB to the `g_iob_committed` list **if it's going to be used by a task waiting for a throttled IOB**. Otherwise, we don't decrement `g_iob_sem.semcount` because it already represents the amount of free resources (in this case, less or equal to zero, if a task is waiting for a non-throttled IOB).

I simplified the implementation one step further. If an IOB is being freed and if a task is waiting for it, it will be put in the `g_iob_committed` list. If it is going to be committed to a task waiting for a throttled IOB, `committed_thottled` is set.  If set, we decrement it from the `g_iob_sem.semcount` before posting to the `g_throttle_sem`. No decrement occurs in `iob_allocwait`.

This satisfies the corner cases. The same tests were performed under similar results with this new implementation.

**Review by wengzhe — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]
